pykechain/models/widgets/widget.py

```diff
@@ -195,11 +191,68 @@ def update_associations(self, readable_models=None, writable_models=None, **kwar
         :raises APIError: when the associations could not be changed
         :raise IllegalArgumentError: when the list is not of the right type
         """
-        self._client.update_widget_associations(widget=self, readable_models=readable_models,
-                                                writable_models=writable_models, **kwargs)
+        self._client.update_widget_associations(
+            widget=self,
+            readable_models=readable_models,
+            writable_models=writable_models,
+            **kwargs
+        )
+
+    def set_associations(
+            self,
+            readable_models: Optional[List] = None,
+            writable_models: Optional[List] = None,
+            **kwargs
+    ) -> None:
+        """
+        Set associations on this widget.
+
+        This is an absolute list of associations. If you provide No models, than the associations are cleared.
+
+        Alternatively you may use `inputs` or `outputs` as a alias to `readable_models` and `writable_models`
+        respectively.
+
+        :param readable_models: list of property models (of :class:`Property` or property_ids (uuids) that has
```

Comment:
we also have instance level associations. For a property grid you only associate a single instance, no models. Model associations is done only for grid types.

I need to check the backend to understand what is really capable

Reply:
In the backend in KE-chain-2 we also can have part_instance_id for attachmentviewer (signature widgets) and propertygrid.

see the implementation

```python
class WidgetUpdateAssociationsSerializer(WidgetUpdateSerializer):
    """Updates the Associations of widget"""

    id = UUIDField(required=False)
    readable_model_properties_ids = ListField(
        child=UUIDField(allow_null=False), write_only=True, required=False
    )
    writable_model_properties_ids = ListField(
        child=UUIDField(allow_null=False), write_only=True, required=False
    )

    parent_part_instance_id = UUIDField(
        write_only=True, required=False
    )  # for partGrid/filteredGrid
    part_instance_id = UUIDField(
        write_only=True, required=False
    )  # for propertyGrid/attachmentViewer

    class Meta(WidgetUpdateSerializer.Meta):
        model = WidgetUpdateSerializer.Meta.model
        fields = WidgetUpdateSerializer.Meta.fields + (
            "readable_model_properties_ids",
            "writable_model_properties_ids",
            "parent_part_instance_id",
            "part_instance_id",
        )
```